### ISSUE TITLE
BRDF distribution term goes nuts with very low roughness and very high NdotL

### ISSUE THREAD

**Issue by TaaTT4** [OGRECave/ogre-next#189]
I've found a huge discrepancy in distribution term with alpha/roughness very low (~0.001) and NdotL very high (~1.0).
You can easily have a proof of it looking at here: https://www.desmos.com/calculator/ofhoyuu1cj
* alpha is the roughness;
* g1 is the standard GGX distribution term (without PI division, like OGRE does);
* [f](https://github.com/OGRECave/ogre-next/blob/master/Samples/Media/Hlms/Pbs/Any/Main/200.BRDFs_piece_ps.any#L137) and [R](https://github.com/OGRECave/ogre-next/blob/master/Samples/Media/Hlms/Pbs/Any/Main/200.BRDFs_piece_ps.any#L138) are the OGRE variables used in BRDF shader;
* d is the absolute difference between the theoretical value g1 and the value effectively computed by OGRE;

The discrepancy is caused by adding `1e-6f` to the denominator in the R calculation (to prevent division by 0, I guess). I don't know if the same issue happens in the visibility term calculation as well, where a similar approach is taken both for [height-correlated](https://github.com/OGRECave/ogre-next/blob/master/Samples/Media/Hlms/Pbs/Any/Main/200.BRDFs_piece_ps.any#L145) and [non height-correlated](https://github.com/OGRECave/ogre-next/blob/master/Samples/Media/Hlms/Pbs/Any/Main/200.BRDFs_piece_ps.any#L149)) version.

**Comment by darksylinc** [comment]
> to prevent a division by 0, I guess; which can NEVER happen since roughness lower bound is set to 0.001

Mmm this is a very good point! The extra addition can be removed!

**Comment by TaaTT4** [comment]
[This](https://github.com/google/filament/blob/main/shaders/src/brdf.fs#L54) is how Filament computes GGX. Division by PI apart, I guess I'll get the same results removing `1e-6f` (please note that [`saturateMediump(x)` is defined as `x`](https://github.com/google/filament/blob/main/shaders/src/common_math.fs#L18) for non-mobile platforms).

**Comment by darksylinc** [comment]
OK I took a fresh look at you results.

HOLY CRAP THAT IS LARGE DIFFERENCE.

You're right the 1e-6f needs to go, and it is safe to do so in `R = sqR / (f * f + 1e-6f);`

**Comment by TaaTT4** [comment]
> Regarding:
> `float G = 0.5 / (( Lambda_GGXV + Lambda_GGXL + 1e-6f ) * 3.141592654);` I believe there is no such guarantee.
> 
> If NdotL is 0, G becomes infinite.

You're right. At least one of `NdotV` or `NdotL` must be greater than zero. Instead of using an epsilon, Filament took a different approach:
```
#define MIN_N_DOT_V 1e-4

float clampNoV(float NoV) {
    // Neubelt and Pettineo 2013, "Crafting a Next-gen Material Pipeline for The Order: 1886"
    return max(NoV, MIN_N_DOT_V);
}

shading_NoV = clampNoV(dot(shading_normal, shading_view));
```
But I don't know which choice gives _better_ results among using an epsilon vs clamping `NdotV` at the source.

> As for the height-uncorrelated:
> `float G = 1.0 / (( gL * gV + 1e-4f ) * 4 * 3.141592654);`
> 
> IIRC the 1e-4f was a magic value because G was so big it looked 'obviously' wrong.

Marmoset 3 (which is based on a non height-correlated geometry function) uses a different constant at `G` numerator (0.25 instead of 0.5) and uses the non squared roughness (`a` is equal to `perceptualRoughness * perceptualRoughness`) scaled by a half:
```
//geometric / visibility
float k = a * 0.5;
float G_SmithL = LdotN * (1.0 - k) + k;
float G_SmithV = VdotN * (1.0 - k) + k;
float G = 0.25 / ( G_SmithL * G_SmithV );
```
Again, I don't know which approach leads to _better_ results.

**Comment by AllianceDarkSylinc** [comment]
We went over this one before.

Our formula (I will ignore the epsilon and PI):
```
k = pixelData.roughness * pixelData.roughness
float gL = NdotL * (1-k) + k;
float gV = pixelData.NdotV * (1-k) + k;
float G = 1.0 / (gL * gV * 4);
float G = 0.25 / (gL * gV); // This is mathematically equivalent
```

Their formula you posted:
```
k = perceptualRoughness * perceptualRoughness * 0.5;
float G_SmithL = NdotL * (1.0 - k) + k;
float G_SmithV = NdotV * (1.0 - k) + k;
float G = 0.25 / ( G_SmithL * G_SmithV );
```

So the only main difference would appear in how k is calculated.

Now the other two differences is that we divide by PI, which was taking from another term, so it should be harmless... (R / PI) * G = R * (G / PI) and that we have an epsilon to prevent division by 0.

**Comment by darksylinc** [comment]
We went over this one before.

Our formula (I will ignore the epsilon and PI):
```
k = pixelData.roughness * pixelData.roughness
float gL = NdotL * (1-k) + k;
float gV = pixelData.NdotV * (1-k) + k;
float G = 1.0 / (gL * gV * 4);
float G = 0.25 / (gL * gV); // This is mathematically equivalent
```

Their formula you posted:
```
k = perceptualRoughness * perceptualRoughness * 0.5;
float G_SmithL = NdotL * (1.0 - k) + k;
float G_SmithV = NdotV * (1.0 - k) + k;
float G = 0.25 / ( G_SmithL * G_SmithV );
```

So the only main difference would appear in how k is calculated.

Now the other two differences is that we divide by PI, which was taking from another term, so it should be harmless... (R / PI) * G = R * (G / PI) and that we have an epsilon to prevent division by 0.
